services/mlmodel/tflite_cpu/tflite_cpu.go

```diff
@@ -0,0 +1,262 @@
+// Package tflitecpu is meant to implement a tfliteCPU in the ML model service
+package tflitecpu
+
+import (
+	"context"
+	fp "path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/edaniels/golog"
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+
+	"go.viam.com/rdk/config"
+	inf "go.viam.com/rdk/ml/inference"
+	"go.viam.com/rdk/ml/inference/tflite_metadata"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/services/mlmodel"
+	"go.viam.com/rdk/services/vision"
+	"go.viam.com/rdk/utils"
+)
+
+func init() {
+	registry.RegisterService(vision.Subtype, resource.DefaultServiceModel, registry.Service{
+		Constructor: func(ctx context.Context, deps registry.Dependencies, config config.Service, logger golog.Logger) (interface{}, error) {
+			return newTFLiteCPUModel(ctx, deps, config, logger)
+		},
+	})
+	sModel := resource.NewDefaultModel("tflitecpu")
+	config.RegisterServiceAttributeMapConverter(mlmodel.Subtype, sModel, func(attributes config.AttributeMap) (interface{}, error) {
+		// Read ML model service parameters into a TFLiteConfig
+		var t TFLiteConfig
+		tfParams, err := config.TransformAttributeMapToStruct(&t, attributes)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error getting parameters from config")
+		}
+		params, ok := tfParams.(*TFLiteConfig)
+		if !ok {
+			return nil, utils.NewUnexpectedTypeError(params, tfParams)
+		}
+		return params, nil
+	}, &TFLiteConfig{})
+}
+
+// TFLiteConfig contains the parameters specific to a tflite_cpu implementation
+// of the MLMS (machine learning model service).
+type TFLiteConfig struct {
+	// this should come from the attributes of the tflite_cpu instance of the MLMS
+	ModelPath  string  `json:"model_path"`
+	NumThreads int     `json:"num_threads"`
+	LabelPath  *string `json:"label_path"`
+}
+
+// Model is a struct that implements the TensorflowLite CPU implementation of the MLMS.
+// It includes the configured parameters, model struct, and associated metadata.
+type Model struct {
+	attrs    TFLiteConfig
+	model    *inf.TFLiteStruct
+	metadata *mlmodel.MLMetadata
+}
+
+// newTFLiteCPUModel is a constructor that builds a tflite cpu implementation of the MLMS.
+//nolint:unparam
+func newTFLiteCPUModel(ctx context.Context, deps registry.Dependencies,
+	conf config.Service, logger golog.Logger,
+) (mlmodel.Service, error) {
+	ctx, span := trace.StartSpan(ctx, "service::mlmodel::NewTFLiteCPUModel")
+	defer span.End()
+
+	svcConfig, ok := conf.ConvertedAttributes.(*TFLiteConfig)
+	if !ok {
+		return nil, utils.NewUnexpectedTypeError(svcConfig, conf.ConvertedAttributes)
+	}
+
+	return CreateTFLiteCPUModel(ctx, svcConfig)
+}
+
+// CreateTFLiteCPUModel is a constructor that builds a tflite cpu implementation of the MLMS.
+func CreateTFLiteCPUModel(ctx context.Context, params *TFLiteConfig) (*Model, error) {
```

Comment:
You've split up `newTFLiteCPUModel`, `CreateTFLiteCPUModel`, and `addTFLiteModel` into 3 different functions. What are the differences between all of them? Why not just have one function?

Reply:
Mostly did this for clear error handling, but you're right. I've combined them into one and we still have the clear error handling.